### ISSUE TITLE
Wrong time range when meeting incomplete metrics

### ISSUE THREAD

**Issue by tmyl123** [grafana/grafana#9776]
- What Grafana version are you using?
4.7.0-10031pre1

- What datasource are you using?
Elasticsearch 5.5.1

- What OS are you running grafana on?
Ubuntu 16.04

- What did you do?
Add Now delay/ Time shift to avoid incomplete metrics of data aggregation on graph

- What was the expected result?
Avoid zero/null on the graph

- What happened instead?
No matter what the time range is, the incomplete data always shown at the end of graph

Normal  :  ↓ ↓ ↓  logstash not collect yet, thus show 0 at the end.
![screenshot_2017-11-03_16-00-36](https://user-images.githubusercontent.com/15243955/32364507-35121580-c0b0-11e7-8de7-69ad59070fee.png)

Weired-1 : ↓ ↓ ↓  add Now delay, but the end of the graph still shows 0
![screenshot_2017-11-03_16-10-39](https://user-images.githubusercontent.com/15243955/32364751-975e2098-c0b1-11e7-80e8-8ecb22469343.png)

Weired-2 : ↓ ↓ ↓  when it happens (show 0 at the end), no matter what time range we select, **It Just Happens**
![screenshot_2017-11-03_16-12-45](https://user-images.githubusercontent.com/15243955/32364782-d98124ac-c0b1-11e7-90e1-629576f94ba1.png)

<br>
<br>

- request & response of normal case (every metric show properly):

request
```
{"search_type":"query_then_fetch","ignore_unavailable":true,"index":["wowza_out_tmp-2017.11.03"]}
{"size":0,"query":{"bool":{"filter":[{"range":{"@timestamp":{"gte":"1509696838403","lte":"1509697138403","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"*"}}]}},"aggs":{"2":{"date_histogram":{"interval":"1m","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1509696838403","max":"1509697138403"},"format":"epoch_millis"},"aggs":{"1":{"sum":{"field":"bw_out"}}}}}}
```

response
```
{"responses":[{"took":2,"timed_out":false,"_shards":{"total":2,"successful":2,"failed":0},"hits":{"total":48,"max_score":0.0,"hits":[]},"aggregations":{"2":{"buckets":[{"key_as_string":"1509696780000","key":1509696780000,"doc_count":0,"1":{"value":0.0}},{"key_as_string":"1509696840000","key":1509696840000,"doc_count":9,"1":{"value":5.080000013113022}},{"key_as_string":"1509696900000","key":1509696900000,"doc_count":9,"1":{"value":5.00999990105629}},{"key_as_string":"1509696960000","key":1509696960000,"doc_count":9,"1":{"value":6.060000061988831}},{"key_as_string":"1509697020000","key":1509697020000,"doc_count":10,"1":{"value":5.859999895095825}},{"key_as_string":"1509697080000","key":1509697080000,"doc_count":11,"1":{"value":4.689999967813492}}]}},"status":200}]}
```

<br>
<br>

- request & response of incomplete metrics (show 0 at end of the graph):

request
```
{"search_type":"query_then_fetch","ignore_unavailable":true,"index":["wowza_out_tmp-2017.11.03"]}
{"size":0,"query":{"bool":{"filter":[{"range":{"@timestamp":{"gte":"1509697076732","lte":"1509697376732","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"*"}}]}},"aggs":{"2":{"date_histogram":{"interval":"1m","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1509697076732","max":"1509697376732"},"format":"epoch_millis"},"aggs":{"1":{"sum":{"field":"bw_out"}}}}}}
```

response
```
{"responses":[{"took":2,"timed_out":false,"_shards":{"total":2,"successful":2,"failed":0},"hits":{"total":41,"max_score":0.0,"hits":[]},"aggregations":{"2":{"buckets":[{"key_as_string":"1509697020000","key":1509697020000,"doc_count":0,"1":{"value":0.0}},{"key_as_string":"1509697080000","key":1509697080000,"doc_count":11,"1":{"value":4.689999967813492}},{"key_as_string":"1509697140000","key":1509697140000,"doc_count":9,"1":{"value":4.89999994635582}},{"key_as_string":"1509697200000","key":1509697200000,"doc_count":10,"1":{"value":6.5099999606609344}},{"key_as_string":"1509697260000","key":1509697260000,"doc_count":11,"1":{"value":5.900000095367432}},{"key_as_string":"1509697320000","key":1509697320000,"doc_count":0,"1":{"value":0.0}}]}},"status":200}]}
```

<br>
<br>

- request & response of incomplete metrics & add Now delay (**still** show 0 at end of the graph):

request
```
{"search_type":"query_then_fetch","ignore_unavailable":true,"index":["wowza_out_tmp-2017.11.03"]}
{"size":0,"query":{"bool":{"filter":[{"range":{"@timestamp":{"gte":"1509697219479","lte":"1509697399479","format":"epoch_millis"}}},{"query_string":{"analyze_wildcard":true,"query":"*"}}]}},"aggs":{"2":{"date_histogram":{"interval":"1m","field":"@timestamp","min_doc_count":0,"extended_bounds":{"min":"1509697219479","max":"1509697399479"},"format":"epoch_millis"},"aggs":{"1":{"sum":{"field":"bw_out"}}}}}}
```

response
```
{"responses":[{"took":3,"timed_out":false,"_shards":{"total":2,"successful":2,"failed":0},"hits":{"total":32,"max_score":0.0,"hits":[]},"aggregations":{"2":{"buckets":[{"key_as_string":"1509697200000","key":1509697200000,"doc_count":10,"1":{"value":6.5099999606609344}},{"key_as_string":"1509697260000","key":1509697260000,"doc_count":11,"1":{"value":5.900000095367432}},{"key_as_string":"1509697320000","key":1509697320000,"doc_count":11,"1":{"value":5.759999958798289}},{"key_as_string":"1509697380000","key":1509697380000,"doc_count":0,"1":{"value":0.0}}]}},"status":200}]}
```

---
we notice that when we choose ```avg``` instead of ```sum```, the graph will keep the last time bucket empty (since it's null), which we could totally accept that. but the behavior of elasticsearch is give ```0``` when no metric to sum, and give ```null``` when no metric to avg, I think that's why we are getting 0.

but the thing that really confusing me is that why isn't adding time shift work? what cause that 0 always shown when there should not (metric exactly exist at such time point)?

**Comment by torkelo** [comment]
strange, maybe a bug in Elasticsearch with sum? With the now delay ES should not return a 0 value but seems it is. 

Try changing min doc count to from zero to 1

**Comment by torkelo** [comment]
I tested this I do not get the same behavior when I increase the date histogram interval (set a minimum interval is recommend). or change min doc count to 1

**Comment by tmyl123** [comment]
Wow, that did the trick, make the end of the graph hold last value and don't drop to 0 (like avg do).
but, we found that will have some mis-render.
eg.
when there is no doc, and sum is zero indeed, the graph should have a line a 0.
but if we set min doc count to 1, the line will connect between two metrics.

Origin :  ↓ ↓ ↓ 
![image](https://user-images.githubusercontent.com/15243955/32378130-e88730f8-c0e4-11e7-9a90-b9ee3a93f4e6.png)

<br>

After :  ↓ ↓ ↓ 
![image](https://user-images.githubusercontent.com/15243955/32378254-31d7ad6e-c0e5-11e7-97c5-31caf2b3b534.png)


the date interval, that did it, too, the higher the interval, the lower chance we see zero.
seems like It's because the interval bucket? when we set interval at 2m, we'll get zero at every even minute, and graph will come up when logstash send value to elasticsearch within this 2 minute bucket.

and if we set to 10m, we'll get 0 at every 10, 20, 30 ..minute.

but we have logstash http_poller every minute, which mean, change interval to 2m, will cause the value is not also sum by our desired query, but also the time bucket (double the metric in this case) 

So looks like set min doc count to 1 at date histogram is our best chance, right?
